File: typescript/test/client.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { createServer, type IncomingMessage, type ServerResponse } from "node:http";
import type { AddressInfo } from "node:net";
import { AmdyClient, UnauthorizedError, NotFoundError, AmdyError } from "../src/client";

/** Start a local HTTP fixture server; returns base URL and captured requests. */
async function startFixture(
  handler: (req: IncomingMessage, res: ServerResponse, body: string) => void
) {
  const requests: { method: string; url: string; headers: any; body: string }[] = [];
  const server = createServer((req, res) => {
    let data = "";
    req.on("data", (c) => (data += c));
    req.on("end", () => {
      requests.push({ method: req.method!, url: req.url!, headers: req.headers, body: data });
      handler(req, res, data);
    });
  });
  await new Promise<void>((r) => server.listen(0, "127.0.0.1", r));
  const port = (server.address() as AddressInfo).port;
  return {
    baseUrl: `http://127.0.0.1:${port}`,
    requests,
    close: () => new Promise<void>((r) => server.close(() => r())),
  };
}

const KEY = "amd_live_testkey123";

test("health() GETs /api/health with no auth header", async () => {
  const fx = await startFixture((req, res) => {
    res.setHeader("content-type", "application/json");
    res.end(JSON.stringify({ status: "ok" }));
  });
  try {
    const client = new AmdyClient({ apiKey: KEY, baseUrl: fx.baseUrl });
    const health = await client.health();
    assert.deepEqual(health, { status: "ok" });
    const r = fx.requests[0];
    assert.equal(r.method, "GET");
    assert.equal(r.url, "/api/health");
    assert.equal(r.headers.authorization, undefined);
    assert.equal(r.headers["x-api-key"], undefined);
  } finally {
    await fx.close();
  }
});

test("getConfig() sends Authorization: Bearer by default", async () => {
  const fx = await startFixture((req, res) => {
    res.setHeader("content-type", "application/json");
    res.end(
      JSON.stringify({
        clientId: 1042,
        detectionSensitivity: 3,
        maxDetectionMs: 8000,
        updatedAt: "2026-08-28T14:02:11.000Z",
      })
    );
  });
  try {
    const client = new AmdyClient({ apiKey: KEY, baseUrl: fx.baseUrl });
    const cfg = await client.getConfig();
    assert.equal(cfg.clientId, 1042);
    assert.equal(cfg.detectionSensitivity, 3);
    assert.equal(cfg.maxDetectionMs, 8000);
    const r = fx.requests[0];
    assert.equal(r.method, "GET");
    assert.equal(r.url, "/api/v1/config");
    assert.equal(r.headers.authorization, `Bearer ${KEY}`);
  } finally {
    await fx.close();
  }
});

test("authStyle header sends X-API-Key instead of Authorization", async () => {
  const fx = await startFixture((req, res) => {
    res.setHeader("content-type", "application/json");
    res.end(JSON.stringify({ ok: true }));
  });
  try {
    const client = new AmdyClient({ apiKey: KEY, baseUrl: fx.baseUrl, authStyle: "header" });
    await client.listIps();
    const r = fx.requests[0];
    assert.equal(r.url, "/api/v1/ips");
    assert.equal(r.headers["x-api-key"], KEY);
    assert.equal(r.headers.authorization, undefined);
  } finally {
    await fx.close();
  }
});

test("registerIp() POSTs to the right path with no request body", async () => {
  const fx = await startFixture((req, res) => {
    res.setHeader("content-type", "application/json");
    res.end(JSON.stringify({ registered: true }));
  });
  try {
    const client = new AmdyClient({ apiKey: KEY, baseUrl: fx.baseUrl });
    const result = await client.registerIp();
    assert.deepEqual(result, { registered: true });
    const r = fx.requests[0];
    assert.equal(r.method, "POST");
    assert.equal(r.url, "/api/v1/ips/register");
    assert.equal(r.headers["content-type"], undefined);
    assert.equal(r.body, "");
  } finally {
    await fx.close();
  }
});

test("401 maps to UnauthorizedError with server message", async () => {
  const fx = await startFixture((req, res) => {
    res.statusCode = 401;
    res.setHeader("content-type", "application/json");
    res.end(JSON.stringify({ error: "Missing Authorization: Bearer <api_key>" }));
  });
  try {
    const client = new AmdyClient({ apiKey: KEY, baseUrl: fx.baseUrl });
    await assert.rejects(client.getClientSettings(), (err: unknown) => {
      assert.ok(err instanceof UnauthorizedError);
      assert.equal((err as UnauthorizedError).message, "Missing Authorization: Bearer <api_key>");
      assert.equal((err as UnauthorizedError).status, 401);
      return true;
    });
  } finally {
    await fx.close();
  }
});

test("404 maps to NotFoundError", async () => {
  const fx = await startFixture((req, res) => {
    res.statusCode = 404;
    res.setHeader("content-type", "application/json");
    res.end(JSON.stringify({ error: "Client not found" }));
  });
  try {
    const client = new AmdyClient({ apiKey: KEY, baseUrl: fx.baseUrl });
    await assert.rejects(client.getConfig(), (err: unknown) => {
      assert.ok(err instanceof NotFoundError);
      assert.equal((err as NotFoundError).status, 404);
      return true;
    });
  } finally {
    await fx.close();
  }
});

test("other error statuses map to AmdyError with status", async () => {
  const fx = await startFixture((req, res) => {
    res.statusCode = 500;
    res.setHeader("content-type", "application/json");
    res.end(JSON.stringify({ error: "boom" }));
  });
  try {
    const client = new AmdyClient({ apiKey: KEY, baseUrl: fx.baseUrl });
    await assert.rejects(client.listIps(), (err: unknown) => {
      assert.ok(err instanceof AmdyError);
      assert.equal((err as AmdyError).status, 500);
      assert.equal((err as AmdyError).message, "boom");
      return true;
    });
  } finally {
    await fx.close();
  }
});

test("missing apiKey is rejected at construction", () => {
  assert.throws(() => new AmdyClient({ apiKey: "" }), AmdyError);
});

test("trailing slashes in baseUrl are normalized", async () => {
  const fx = await startFixture((req, res) => {
    res.setHeader("content-type", "application/json");
    res.end(JSON.stringify({}));
  });
  try {
    const client = new AmdyClient({ apiKey: KEY, baseUrl: `${fx.baseUrl}/` });
    await client.health();
    assert.equal(fx.requests[0].url, "/api/health");
  } finally {
    await fx.close();
  }
});

File: ruby/test/client_test.rb
# frozen_string_literal: true

require_relative "test_helper"
require "minitest/autorun"

class ClientTest < Minitest::Test
  def setup
    @server = FixtureServer.new
    until @server.started?
      raise "fixture failed to start" unless @server.started? || sleep(0.05)
    end
  end

  def teardown
    @server.stop
  end

  def test_health_returns_parsed_json_without_auth_header
    result = @server.client.health
    assert_equal({ "status" => "ok" }, result)
    last = @server.requests.last
    assert_equal "GET", last[:method]
    assert_equal "/api/health", last[:path]
    assert_nil last[:authorization], "health must not send auth header"
  end

  def test_config_returns_hash_values
    config = @server.client.config
    assert_equal 1042, config["clientId"]
    assert_equal 3, config["detectionSensitivity"]
    assert_equal 8000, config["maxDetectionMs"]
    assert_equal "2026-08-28T14:02:11.000Z", config["updatedAt"]
    assert_equal "Bearer amd_live_test_key", @server.requests.last[:authorization]
  end

  def test_client_settings_passthrough
    settings = @server.client.client_settings
    assert_equal({ "webhookUrl" => "https://example.test/hook", "timezone" => "UTC" }, settings)
  end

  def test_ips_passthrough
    ips = @server.client.ips
    assert_equal({ "ips" => ["203.0.113.7", "198.51.100.9"] }, ips)
  end

  def test_register_ip_posts_to_correct_path_without_body
    result = @server.client.register_ip
    assert_equal({ "registered" => true }, result)
    last = @server.requests.last
    assert_equal "POST", last[:method]
    assert_equal "/api/v1/ips/register", last[:path]
    # Net::HTTP stamps a default content type on POST; the important part is the empty body.
    assert_equal "", last[:body].to_s
  end

  def test_401_raises_auth_error_with_server_message
    @server.stub_status("/api/v1/config", status: 401, body: { "error" => "Invalid API key" })
    error = assert_raises(Amdy::AuthError) { @server.client.config }
    assert_equal 401, error.status
    assert_equal "Invalid API key", error.message
  end

  def test_404_raises_not_found_error
    @server.stub_status("/api/v1/config", status: 404, body: { "error" => "Client not found" })
    error = assert_raises(Amdy::NotFoundError) { @server.client.config }
    assert_equal 404, error.status
    assert_equal "Client not found", error.message
  end

  def test_other_error_raises_api_error_and_extracts_message
    @server.stub_status("/api/v1/config", status: 429, body: { "error" => "Rate limit exceeded" })
    error = assert_raises(Amdy::ApiError) { @server.client.config }
    assert_equal 429, error.status
    assert_equal "Rate limit exceeded", error.message
    refute_kind_of Amdy::AuthError, error
    refute_kind_of Amdy::NotFoundError, error
  end

  def test_error_without_message_falls_back_to_status_message
    @server.stub_status("/api/v1/config", status: 500, body: {})
    error = assert_raises(Amdy::ApiError) { @server.client.config }
    assert_equal "HTTP 500", error.message
  end

  def test_connection_error_on_unreachable_server
    # Port 1 on localhost is unassigned; connection is refused.
    client = Amdy::Client.new(api_key: "amd_live_test_key", base_url: "http://127.0.0.1:1", open_timeout: 1, read_timeout: 1)
    assert_raises(Amdy::ConnectionError) { client.health }
  end

  def test_client_requires_api_key
    assert_raises(ArgumentError) { Amdy::Client.new(api_key: nil) }
    assert_raises(ArgumentError) { Amdy::Client.new(api_key: "") }
  end
end
